File: src/Visitor.py
from collections import defaultdict
from itertools import count
from typing import List, Mapping, Tuple

from gen.grammarizerParser import grammarizerParser as Parser
from gen.grammarizerVisitor import grammarizerVisitor

from .Node import BranchingNode, Node, SequenceNode, ValueNode


class Visitor(grammarizerVisitor):
    rules_map: Mapping[str, List[Node]] = defaultdict(list)
    counter = count()

    def visitGrammarFile(self, ctx: Parser.GrammarFileContext) -> Mapping[str, List[Node]]:
        for rule_ctx in ctx.rule_():
            name, node = self.visitRule(rule_ctx)
            self.rules_map[name].append(node)
        return self.rules_map

    def visitRule(self, ctx: Parser.RuleContext) -> Tuple[str, Node]:
        node = self.visitBody(ctx.body())
        return ctx.GRAMMAR_RULE_NAME().getText(), node

    def visitBody(self, ctx: Parser.BodyContext) -> BranchingNode:
        alternatives = [self.visitAlternative(c) for c in ctx.alternative()]
        return BranchingNode(alternatives=alternatives)

    def visitAlternative(self, ctx: Parser.AlternativeContext) -> SequenceNode:
        sub_rules = [self.visitUnit(c) for c in ctx.unit()]
        return SequenceNode(sub_rules=sub_rules)

    def visitUnit(self, ctx: Parser.UnitContext) -> Node:
        multiplied_node = (self.visitPrimitive(ctx.primitive())
                           if ctx.primitive()
                           else self.visitGrouping(ctx.grouping()))
        if not ctx.MULTIPLICITY_OP():
            return multiplied_node
        multiplier = ctx.MULTIPLICITY_OP().getText()
        if multiplier == '?':
            alternatives = [multiplied_node, ValueNode(value='')]
            return BranchingNode(alternatives=alternatives)
        aux_rule_name = f"$aux_rule{next(self.counter)}"
        aux_rule_node = ValueNode(value=aux_rule_name)
        aux_rule_alternative = SequenceNode(sub_rules=[multiplied_node, aux_rule_node])
        aux_rule_tree = BranchingNode(alternatives=[multiplied_node, aux_rule_alternative])
        self.rules_map[aux_rule_name].append(aux_rule_tree)
        if multiplier == '*':
            return BranchingNode(alternatives=[aux_rule_node, ValueNode(value='')])
        if multiplier == '+':
            return aux_rule_node
        raise ValueError(f"unhandled multiplier {multiplier}")

    def visitGrouping(self, ctx: Parser.GroupingContext) -> Node:
        return self.visitBody(ctx.body())

    def visitPrimitive(self, ctx: Parser.PrimitiveContext) -> ValueNode:
        return ValueNode(value=ctx.getText())

File: src/Visitor.js
import { range } from "./Itertools.js";

import DefaultMap from "./DefaultMap.js";
import EBNFVisitor from "../gen/EBNFVisitor.js";
import { BranchingNode, SequenceNode, ValueNode } from "./Node.js";


export default class Visitor extends EBNFVisitor {

    constructor() {
        super();
        this.rules_map = new DefaultMap(() => []);
        this.counter = range();
    }

    visitGrammarFile(ctx) {
        for (let rule_ctx of ctx.rule_()) {
            let [name, node] = this.visitRule(rule_ctx);
            this.rules_map.get(name).push(node);
        }
        return this.rules_map;
    }

    visitRule(ctx) {
        let node = this.visitBody(ctx.body());
        return [ctx.GRAMMAR_RULE_NAME().getText(), node];
    }

    visitBody(ctx) {
        let alternatives = ctx.alternative().map((c) => this.visitAlternative(c));
        return new BranchingNode(alternatives);
    }

    visitAlternative(ctx) {
        let sub_rules = ctx.unit().map((c) => this.visitUnit(c));
        return new SequenceNode(sub_rules);
    }

    visitUnit(ctx) {
        let multiplied_node = ctx.primitive() ?
            this.visitPrimitive(ctx.primitive()) :
            this.visitGrouping(ctx.grouping());

        if (!ctx.MULTIPLICITY_OP()) {
            return multiplied_node;
        }

        let multiplier = ctx.MULTIPLICITY_OP().getText();
        if (multiplier === '?') {
            let alternatives = [multiplied_node, new ValueNode('')];
            return new BranchingNode(alternatives);
        }

        let aux_rule_name = `$aux_rule${this.counter.next().value}`;
        let aux_rule_node = new ValueNode(aux_rule_name);
        let aux_rule_alternative = new SequenceNode([multiplied_node, aux_rule_node]);
        let aux_rule_tree = new BranchingNode([multiplied_node, aux_rule_alternative]);
        this.rules_map.get(aux_rule_name).push(aux_rule_tree);
        if (multiplier === '*') {
            return new BranchingNode([aux_rule_node, new ValueNode('')]);
        }
        if (multiplier === '+') {
            return aux_rule_node;
        }
        throw new Error('Unsupported multiplier');
    }

    visitGrouping(ctx) {
        return this.visitBody(ctx.body());
    }

    visitPrimitive(ctx) {
        return new ValueNode(ctx.getText());
    }
}
